rewindable-index/src/RewindableIndex/Storable.hs

```diff
@@ -231,10 +250,42 @@ getEvents
   -> StorableMonad h [StorableEvent h]
 getEvents s = do
   memoryEs <- getMemoryEvents (s ^. storage)
-              & V.freeze <&> V.toList
+              & V.freeze <&> foldEvents . V.toList
   diskEs   <- getStoredEvents (s ^. handle)
   pure $ diskEs ++ memoryEs
 
+{- This function is used to add a checkpoint to the in-memory part of event indexers.
+   You can use checkpoints to add synthetic events to the memory buffer. Filling the
+   memory buffer will flush the real events to disk faster than not using checkpoints.
+
+   This is useful if the number of events is small and you run the risk of never
+   storing them on-disk. When that happens, if the indexer is restarted it will
+   resume from the Genesis block, which is something that is not an expected
+   behaviour.
+
+   Possible future uses of checkpoints:
+
+   * We currently assume that the indexer has processed blocks up to the latest event
+     processed. In case there are very few events this assumption introduces a big
+     error. We may eventually use checkpoints to signal to the indexer that the latest
+     block is more recent than the latest event processed.
+
+   * Currently synthetic events cannot be stored in the persistent database. We may
+     decide to offer this option to implementers if there is interest. This would
+     allow for faster resumes if the last stored event is a lot older than the
+     currently processed block (similar to the in-memory description of checkpoints).
+-}
+checkpoint
```

Comment:
Shouldn't we have a function on top of `checkpoint` and `insert` that could process any `SyntheticEvent`

Reply:
I am not sure we want to couple indexers to a Cardano type. And if we make this implementation for any kind of type, then what we write is basically a `map` function, which I think is better kept outside of the indexer.

I am debating if we need to have a `filter` function within the indexer for similar reasons. The upside of a `filter` function would be that we can make the use of `checkpoint` automatic (checkpoint whatever the filter throws away).

Reply:
Yes, `filter`/`processBlock` (or whatever would be its name) would respond to the same problem. Maybe an `IndexerInput` type family could answer these needs without being Cardano-specific, but it may be hard to type-check if we use it.